src/librustc_mir/dataflow/generic/tests.rs

```diff
@@ -0,0 +1,297 @@
+use rustc::mir::{self, Location, BasicBlock};
+use rustc::ty;
+use rustc_index::bit_set::BitSet;
+use rustc_index::vec::IndexVec;
+
+use crate::dataflow::BottomValue;
+use super::*;
+
+#[test]
+fn cursor_seek() {
+    let body = dummy_body();
+    let body = &body;
+    let analysis = Dummy { body };
+
+    let mut cursor = Results {
+        entry_sets: analysis.entry_sets(),
+        analysis,
+    }.into_cursor(body);
+
+    // Sanity check: the dummy call return effect is unique and actually being applied.
+
+    let call_terminator_loc = Location { block: BasicBlock::from_usize(2), statement_index: 2 };
+    assert!(is_call_terminator_non_diverging(body, call_terminator_loc));
+
+    let call_return_effect =
+        SeekTarget::AfterAssumeCallReturns(call_terminator_loc).effect(body).unwrap();
+    assert_ne!(call_return_effect, SeekTarget::After(call_terminator_loc).effect(body).unwrap());
+
+    cursor.seek_after(call_terminator_loc);
+    assert!(!cursor.get().contains(call_return_effect));
+    cursor.seek_after_assume_call_returns(call_terminator_loc);
+    assert!(cursor.get().contains(call_return_effect));
+
+    let every_target = || body
+        .basic_blocks()
+        .iter_enumerated()
+        .flat_map(|(bb, _)| SeekTarget::iter_in_block(body, bb));
+
+    let mut seek_to_target = |targ| {
+        use SeekTarget::*;
+
+        match targ {
+            BlockStart(block) => cursor.seek_to_block_start(block),
+            Before(loc) => cursor.seek_before(loc),
+            After(loc) => cursor.seek_after(loc),
+            AfterAssumeCallReturns(loc) => cursor.seek_after_assume_call_returns(loc),
+        }
+
+        assert_eq!(cursor.get(), &cursor.analysis().expected_state_at(targ));
+    };
+
+    // Seek *to* every possible `SeekTarget` *from* every possible `SeekTarget`.
+    //
+    // By resetting the cursor to `from` each time it changes, we end up checking some edges twice.
+    // What we really want is an Eulerian cycle for the complete digraph over all possible
+    // `SeekTarget`s, but it's not worth spending the time to compute it.
+    for from in every_target() {
+        seek_to_target(from);
+
+        for to in every_target() {
+            seek_to_target(to);
+            seek_to_target(from);
+        }
+    }
+}
+
+const BASIC_BLOCK_OFFSET: usize = 100;
+
+#[derive(Clone, Copy, Debug, PartialEq, Eq)]
+enum SeekTarget {
+    BlockStart(BasicBlock),
+    Before(Location),
+    After(Location),
+    AfterAssumeCallReturns(Location),
+}
+
+impl SeekTarget {
+    fn block(&self) -> BasicBlock {
+        use SeekTarget::*;
+
+        match *self {
+            BlockStart(block) => block,
+            Before(loc) | After(loc) | AfterAssumeCallReturns(loc) => loc.block,
+        }
+    }
+
+    /// Returns the largest index (not including the basic block index) that should be set when
```

Comment:
I'm trying to understand the docs here, and this method + test-harness in general.

 * I have inferred that the indexes you are talking about are points before and after statements: { } S1 { }, { } S2 { }, ... etc. (At first I wasn't sure why you were distinguishing the point after S1 from the point before S2, but soon I realized that you are identifying points in the whole MIR, not just within a given basic block, and therefore in general you have to deal with control-flow splits and merges.)

 * When you say "that should be set", where are these indices being "*set*"? Is that referring to the `Dummy` that is storing the indices in a set?

 * Why is this called `effect` ? Because it is describing the "effect" of calling `seek` with a given `SeekTarget`? Or just that it describes the effect we want to compute on the bitsets for `Dummy`?

(I suppose part of my problem here is that I started reading this file from top to bottom, and it seems like the important explanations are further down, with `Dummy` and its related impls.)

Reply:
I'll start from first principles since I didn't do a good job of explaining the motivation for this.

The ultimate goal of this is to test the logic that updates the internal state of a `dataflow::ResultsCursor` (this is where indices should be "set") by seeking *to* every possible target (before or after a `Location`) *from* every possible target. For this test to be exhaustive, we must define a dataflow analysis such that every possible effect (before, after and call-return) is unique for every statement and terminator in a MIR body. This ensures that the expected state at every possible target is unique over the entire MIR body. There's no need to actually find the fixpoint for such an analysis to test `ResultsCursor`, so the actual CFG is inconsequential.

The actual implementation does not quite live up to this ideal. Instead it takes advantage of the fact that we never iterate to fixpoint by storing the index of each basic block in its entry set. The effect of each statement in our dummy analysis is only "unique" when considered in conjunction with the index that appears in the entry set.

I could reimplement this test it to be more like the first paragraph. Since I need to map before/after/call-return effects for each statement and terminator to a unique index, I'm afraid that it will always be a bit opaque. I could also keep it roughly as-is with an edited version of this comment as documentation, or remove it entirely.